### PR TITLE
Switch from update to patch for VolumeSnapshot and Content

### LINKED ISSUE #368: Use patch instead of update
To reduce the number of update failures due to resources being out of date, it is recommended to use patch instead of update. For updates where it's important that no other field in the object got modified, then you can include the object version in the patch.

### PR DESCRIPTION
**Pull request by ggriffiths**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Switches to use patch instead of update

**Which issue(s) this PR fixes**:
Fixes #368 

**Special notes for your reviewer**:
n/a

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by ggriffiths** [comment]
Discovered that stategicmergepatch is not supported for CRs. Working now on an alternative

File: go.mod [hunk 1]
```diff
@@ -4,6 +4,7 @@ go 1.15
 
 require (
 	github.com/container-storage-interface/spec v1.3.0
+	github.com/evanphx/json-patch v4.9.0+incompatible
```
Comment (ggriffiths) [inline go.mod:7]: Used in kubernetes/kubernetes:
https://github.com/kubernetes/kubernetes/blob/master/go.mod#L42

File: pkg/common-controller/snapshot_create_test.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-var timeNow = time.Now()
+var timeNow = time.Date(2021, time.January, 1, 0, 0, 0, 0, time.Local)
```
Comment (ggriffiths) [inline pkg/common-controller/snapshot_create_test.go:29]: Was running into unit test issues without this.
```
        -                       CreationTime:                   s"2021-04-01 17:45:55.275503069 -0700 PDT m=+0.021010896",
        +                       CreationTime:                   s"2021-04-01 17:45:55 -0700 PDT",
```

The patching behavior seems to treat timestamps with less precision.

Comment (xing-yang) [reply pkg/common-controller/snapshot_create_test.go:29]: hmm, interesting

Comment (ggriffiths) [reply pkg/common-controller/snapshot_create_test.go:29]: I think the precision is only for unit tests - I just tried on a cluster and the `creationTime` seems to not have the fractional seconds above.

```
  status:
    boundVolumeSnapshotContentName: snapcontent-6ac333b4-25c5-49d9-a55e-5a7cdc8db7f3
    creationTime: "2021-04-02T01:03:11Z"
    readyToUse: true
```

Comment (ggriffiths) [reply pkg/common-controller/snapshot_create_test.go]: Changed this to be `time.Now().Round(time.Second)` instead.

File: deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml [hunk 3]
```diff
@@ -35,10 +35,10 @@ rules:
     verbs: ["get", "list", "watch"]
   - apiGroups: ["snapshot.storage.k8s.io"]
     resources: ["volumesnapshotcontents"]
-    verbs: ["create", "get", "list", "watch", "update", "delete"]
+    verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
```
Comment (xing-yang) [inline deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml:38]: Do we still need "update"?

File: deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml [hunk 4]
```diff
@@ -35,10 +35,10 @@ rules:
...
   - apiGroups: ["snapshot.storage.k8s.io"]
     resources: ["volumesnapshotcontents/status"]
-    verbs: ["update"]
+    verbs: ["update", "patch"]
```
Comment (xing-yang) [inline deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml:41]: Same question about "update"?

File: deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml [hunk 5]
```diff
@@ -34,13 +34,16 @@ rules:
     verbs: ["get", "list", "watch"]
   - apiGroups: ["snapshot.storage.k8s.io"]
     resources: ["volumesnapshotcontents"]
-    verbs: ["create", "get", "list", "watch", "update", "delete"]
+    verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
```
Comment (xing-yang) [inline deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml:37]: Still need "update"?

File: deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml [hunk 6]
```diff
@@ -34,13 +34,16 @@ rules:
...
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents/status"]
+    verbs: ["patch"]
   - apiGroups: ["snapshot.storage.k8s.io"]
     resources: ["volumesnapshots"]
-    verbs: ["get", "list", "watch", "update"]
+    verbs: ["get", "list", "watch", "update", "patch"]
```
Comment (xing-yang) [inline deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml:43]: Still need "update"?

File: deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml [hunk 7]
```diff
@@ -34,13 +34,16 @@ rules:
...
   - apiGroups: ["snapshot.storage.k8s.io"]
     resources: ["volumesnapshots/status"]
-    verbs: ["update"]
+    verbs: ["update", "patch"]
```
Comment (xing-yang) [inline deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml:46]: Still need "update"?

File: pkg/utils/patch_test.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+			updatedSnapshot: &crdv1.VolumeSnapshot{
+				Status: &crdv1.VolumeSnapshotStatus{
+					BoundVolumeSnapshotContentName: nil,
+					CreationTime:                   nil,
+					ReadyToUse:                     nil,
+					RestoreSize:                    nil,
+					Error:                          nil,
+				},
+			},
+			expectedPatch: `{"status":{"boundVolumeSnapshotContentName":null}}`,
```
Comment (xing-yang) [inline pkg/utils/patch_test.go:56]: Should it be "nil"?

File: pkg/utils/patch_test.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+				},
+			},
+			updatedSnapshotContent: &crdv1.VolumeSnapshotContent{
+				Status: &crdv1.VolumeSnapshotContentStatus{
+					SnapshotHandle: nil,
+					CreationTime:   nil,
+					ReadyToUse:     nil,
+					RestoreSize:    nil,
+					Error:          nil,
+				},
```
Comment (xing-yang) [inline pkg/utils/patch_test.go:146]: Add some tests where the updated status has non-nil CreationTime, ReadyToUse, RestoreSize.

File: pkg/utils/patch_test.go [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+					Error:                          nil,
+				},
+			},
+			updatedSnapshot: &crdv1.VolumeSnapshot{
+				Status: &crdv1.VolumeSnapshotStatus{
+					BoundVolumeSnapshotContentName: nil,
+					CreationTime:                   nil,
+					ReadyToUse:                     nil,
+					RestoreSize:                    nil,
+					Error:                          nil,
```
Comment (xing-yang) [inline pkg/utils/patch_test.go:53]: Add some tests where the updated status has non-nil CreationTime, ReadyToUse, RestoreSize.

File: deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml (hunk 3 above)
Comment (ggriffiths) [reply deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml:38]: Removed these update rule.

File: pkg/utils/patch_test.go (hunk 8 above)
Comment (ggriffiths) [reply pkg/utils/patch_test.go:56]: The way merge patch works is that it accepts a json object with all of the fields that are being changed. `null` json value marshals to `nil` in Go.

the createSnapshotPatch/createContentPatch funcs create json objects like this.

File: pkg/utils/patch_test.go (hunk 10 above)
Comment (ggriffiths) [reply pkg/utils/patch_test.go:53]: Added these tests

File: pkg/utils/patch_test.go (hunk 9 above)
Comment (ggriffiths) [reply pkg/utils/patch_test.go:146]: Added these tests

**Comment by ggriffiths** [comment]
Addressed comments, ready for re-review @xing-yang

File: pkg/sidecar-controller/snapshot_controller.go [hunk 11]
```diff
@@ -370,13 +370,14 @@ func (ctrl *csiSnapshotSideCarController) clearVolumeContentStatus(
...
-		content.Status.RestoreSize = nil
+	contentClone := content.DeepCopy()
+	if contentClone.Status != nil {
+		contentClone.Status.SnapshotHandle = nil
+		contentClone.Status.ReadyToUse = nil
+		contentClone.Status.CreationTime = nil
+		contentClone.Status.RestoreSize = nil
 	}
-	newContent, err := ctrl.clientset.SnapshotV1().VolumeSnapshotContents().UpdateStatus(context.TODO(), content, metav1.UpdateOptions{})
+	newContent, err := utils.PatchVolumeSnapshotContent(content, contentClone, ctrl.clientset, "status")
```
Comment (xing-yang) [inline pkg/sidecar-controller/snapshot_controller.go:380]: This string "status" is used in multiple places.  Can you make it a constant?

File: pkg/utils/patch_test.go (hunk 8 above)
Comment (xing-yang) [reply pkg/utils/patch_test.go:61]: Sure.  Can you add this as comments here in the code?

File: pkg/utils/patch.go [hunk 12]
```diff
@@ -42,12 +42,12 @@ func PatchVolumeSnapshot(
...
 	}
 
-	newSnapshot, err := client.SnapshotV1().VolumeSnapshots(updatedSnapshot.Namespace).Patch(context.TODO(), existingSnapshot.Name, types.MergePatchType, patch, metav1.PatchOptions{}, subresources...)
+	_, err = client.SnapshotV1().VolumeSnapshots(updatedSnapshot.Namespace).Patch(context.TODO(), existingSnapshot.Name, types.MergePatchType, patch, metav1.PatchOptions{}, subresources...)
 	if err != nil {
 		return updatedSnapshot, err
 	}
 
-	return newSnapshot, nil
+	return updatedSnapshot, nil
```
Comment (xing-yang) [inline pkg/utils/patch.go:50]: I thought "newSnapshot" is the latest one returned from API server?  Why returning "updatedSnapshot"?

Comment (ggriffiths) [reply pkg/utils/patch.go:50]: this was a test commit - I was trying to see if this reduced the 'modified object' error count.

Comment (ggriffiths) [reply pkg/utils/patch.go:50]: Reverting the above change as it did not lower the error count.

Comment (xing-yang) [reply pkg/utils/patch.go]: Can you take a look of the logs and see if you can tell why we got the "modified object" error?

Comment (ggriffiths) [reply pkg/utils/patch.go]: Asked SIG API machinery per your suggestion to try and find out. I'll also check the logs to see under what scenarios we face this. The error counts are fairly consistent.

**Comment by ggriffiths** [comment]
Still investigating ways to lower the error count. merge patch alone with this implementation does not seem to lower the error count.
/hold

**Comment by ggriffiths** [comment]
Findings so far:
1. Strategicmergepatch is not supported for CRDs. Moving to `mergepatch` via `json-patch` does not lower error count (~250 error count)
2. Removing `addResourceVersion` from implementation does not lower error count (error count: 240)
3. Performing a "Get" before every "Patch" lowered error count to 24 total, but is expensive.

**Comment by ggriffiths** [comment]
As this PR did not actually decrease the "object has been modified" error count, I am closing this for now.

I still plan to investigate how this "object has been modified" problem can be solved.
